### PR TITLE
[SMALLFIX] Put the lineage master and worker behind the flag

### PR DESCRIPTION
**Pull request by yupeng9**

### CODE REVIEW

File: servers/src/main/java/tachyon/master/TachyonMaster.java [hunk 1]
```diff
@@ -170,7 +170,10 @@ protected TachyonMaster() {
       mBlockMaster = new BlockMaster(mBlockMasterJournal);
       mFileSystemMaster = new FileSystemMaster(mBlockMaster, mFileSystemMasterJournal);
       mRawTableMaster = new RawTableMaster(mFileSystemMaster, mRawTableMasterJournal);
-      mLineageMaster = new LineageMaster(mFileSystemMaster, mLineageMasterJournal);
+      boolean enableLineage = MasterContext.getConf().getBoolean(Constants.USER_LINEAGE_ENABLED);
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: `lineageEnabled`?

File: servers/src/main/java/tachyon/master/TachyonMaster.java [hunk 2]
```diff
@@ -279,7 +282,9 @@ protected void startMasters(boolean isLeader) {
       mBlockMaster.start(isLeader);
       mFileSystemMaster.start(isLeader);
       mRawTableMaster.start(isLeader);
-      mLineageMaster.start(isLeader);
+      if (mLineageMaster != null) {
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: `lineageEnabled`?

File: servers/src/main/java/tachyon/master/TachyonMaster.java [hunk 3]
```diff
@@ -289,7 +294,9 @@ protected void startMasters(boolean isLeader) {
 
   protected void stopMasters() {
     try {
-      mLineageMaster.stop();
+      if (mLineageMaster != null) {
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: `lineageEnabled`?

File: servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java [hunk 4]
```diff
@@ -253,6 +256,17 @@ public synchronized void streamToJournalCheckpoint(JournalOutputStream outputStr
...
   /**
+   * Checks if there's an output file with given file id
+   *
+   * @param fileId the file id
+   * @return true if there's a lineage in the store that has the output file of the given id, false
+   *         otherwise
+   */
+  public boolean hasOutputFile(long fileId) {
+    return mOutputFileIndex.containsKey(fileId);
+  }
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java]: Is this method used?

File: servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java [hunk 5]
```diff
@@ -195,10 +195,13 @@ public synchronized Lineage getLineage(long lineageId) {
...
    *
    * @param fileId the file id
-   * @return the lineage containing the output file
+   * @return the lineage containing the output file, null if no lineage outputs the given file
    */
   public synchronized Lineage reportLostFile(long fileId) {
     Lineage lineage = mOutputFileIndex.get(fileId);
+    if (lineage == null) {
+      return null;
+    }
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java]: When would this happen? should an error be thrown?

File: servers/src/main/java/tachyon/master/TachyonMaster.java (hunk 1 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: done

File: servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java (hunk 5 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/lineage/meta/LineageStore.java]: It's possible that the lost file may not associate with any lineage. I'll change the logic to let the caller check it before calling this function.

**Comment by calvinjia** [comment]
LGTM

**Comment by calvinjia** [comment]
@yupeng9 Do you mind changing the variable to no longer be a `user` variable? Since it seems like we do a lot of logic in the servers based on the flag.

**Comment by yupeng9** [comment]
@calvinjia, I set it to be a user variable because it's a feature flag up to the user to turn on/off. But I'm open to other suggestions. What kind of variable that you think is better?

**Pull request merged**